src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -2176,71 +2007,237 @@ lExit: ;
+    // Don't allow generics instantiated over __Canon
     if (pMT->IsSharedByGenericInstantiations()) {
         COMPlusThrow(kNotSupportedException, W("NotSupported_Type"));
     }
 
-    // Never allow the allocation of an unitialized ContextBoundObject derived type, these must always be created with a paired
-    // transparent proxy or the jit will get confused.
+    // Don't allow ref structs
+    if (pMT->IsByRefLike()) {
+        COMPlusThrow(kNotSupportedException, W("NotSupported_ByRefLike"));
+    }
+
+    // Caller passed Nullable<T> but wanted information for T instead?
+    if (fUnwrapNullable && Nullable::IsNullableType(pMT))
+    {
+        pMT = pMT->GetInstantiation()[0].GetMethodTable();
+    }
 
 #ifdef FEATURE_COMINTEROP
-    // Also do not allow allocation of uninitialized RCWs (COM objects).
+    // COM allocation can involve the __ComObject base type (with attached CLSID) or a
+    // VM-implemented [ComImport] class. For GetUninitializedObject, we block all COM types.
+    // For CreateInstance, the flowchart is:
+    //   - For __ComObject,
+    //     .. on Windows, bypass normal newobj logic and use ComClassFactory::CreateInstance.
+    //     .. on non-Windows, treat as a normal class, type has no special handling in VM.
+    //   - For [ComImport] class, treat as a normal class. VM will replace default
+    //     ctor with COM activation logic on supported platforms, else ctor itself will PNSE.
     if (pMT->IsComObjectType())
-        COMPlusThrow(kNotSupportedException, W("NotSupported_ManagedActivation"));
+    {
+        if (fGetUninitializedObject)
+        {
+            COMPlusThrow(kNotSupportedException, W("NotSupported_ManagedActivation"));
+        }
+
+        if (IsComObjectClass(typeHandle))
+        {
+#ifdef FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+            // Need to enter cooperative mode to manipulate OBJECTREFs
+            GCX_COOP();
+            SyncBlock* pSyncBlock = pRuntimeType.Get()->GetSyncBlock();
+            pClassFactory = (void*)pSyncBlock->GetInteropInfo()->GetComClassFactory();
+#endif // FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+
+            if (!pClassFactory)
+            {
+                COMPlusThrow(kInvalidComObjectException, IDS_EE_NO_BACKING_CLASS_FACTORY);
+            }
+
+            fRequiresSpecialComActivationStub = true;
+        }
+    }
 #endif // FEATURE_COMINTEROP
 
-    // If it is a nullable, return the underlying type instead.
-    if (Nullable::IsNullableType(pMT))
-        pMT = pMT->GetInstantiation()[0].GetMethodTable();
+    pMT->EnsureInstanceActive();
 
-    retVal = pMT->Allocate();
+    // All checks passed! Pass parameters back to the caller.
+
+    if (fRequiresSpecialComActivationStub)
+    {
+        // managed sig: ComClassFactory* -> object (via FCALL)
+        // caller understands the -1 sentinel value and substitutes real address
+        *ppfnAllocator = (PCODE)(-1);
```

Comment:
You can get the right pointer here by:

`*ppfnAllocator = CoreLibBinder::GetMethod(METHOD__....)->GetMultiCallableAddrOfCode()`

+ define entry for the `METHOD__` in `corelib.h`.

Reply:
Thanks! I was using `*ppfnAllocator = (PCODE)GetEEFuncEntryPoint(AllocateComObject);` earlier. While everything appeared to work just fine at runtime, the GC started failing asserts when trying to walk back stack frames since it couldn't figure out the corresponding managed method. (I guess the _\_\_me_ variable wasn't set properly?)

Reply:
The runtime needs mapping between methoddesc and entrypoint in number of places. `GetEEFuncEntryPoint` does not populate this mapping and that's why you probably saw the asserts.